### DISCUSSION TITLE
Problems with Keras module

### DISCUSSION THREAD

**Discussion by mlauronen** [microsoft/pyright#6145] [Q&A] [answered]
I am not sure if this is a bug or what but I'm getting 2 errors while accessing Keras module:
- Cannot access member shape
- evaluate is not a known member

**Any idea how to fix ?**

The code works OK when run. 
Please see [the code here](https://i.imgur.com/UdnNFCt.png).

**Comment by erictraut** [comment]
If you would like help, please provide a minimal, self-contained code sample in text form. Diagnosing issues from partial screen shots isn't really practical.

**Comment by mlauronen** [comment]
Well that was the whole script. But here is it in the text form:
```python
import argparse
from pandas import read_csv
from sklearn.model_selection import train_test_split
from keras import Sequential
from keras.layers import Dense
from keras.models import load_model

parser = argparse.ArgumentParser("")
parser.add_argument(
    '-d', '--debug', action='store_true', help='Enable debug mode')
parser.add_argument(
    '-t', '--train', action='store_true', help='Train the model')
args = parser.parse_args()

model_file = 'model.keras'
data = read_csv('data.csv')

X = data.iloc[:, :-1]  # Features in columns except the last column
y = data.iloc[:, -1]   # Last column is the desired output

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)

if args.train:
    model = Sequential([
        Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
        Dense(32, activation='relu'),
        Dense(1)
    ])
    model.compile(optimizer='adam', loss='mean_squared_error', metrics=['mae'])
    model.fit(X_train, y_train, epochs=10, batch_size=32)
    model.save(model_file)
else:
    model = load_model(model_file)

loss, mae = model.evaluate(X_test, y_test)

print(f"Loss: {loss}")
print(f"MAE: {mae}")
```

**Reply by mlauronen** [reply]
btw. How do you add the code in a more formatted way?
Ctrl-E didn't work...

**EDIT:** I figured it out

**Reply by erictraut** [reply]
That's definitely not a minimal example. In the future, please try to pare down your code to remove the parts that are unrelated to your question. In the process of simplifying your code, you can often discover the cause of the problem and answer the question yourself.

You're importing for a bunch of different libraries in this code. Which versions of these libraries do you have installed? Do you have type stubs installed for any of them? The `argparse` library has typeshed stubs, so that's fine. Recent versions of `pandas` have inlined type information, so that's probably fine too. But `sklearn` and `keras` are completely untyped. If you don't have type stubs installed for them, pyright will do its best to infer types, but this is error prone — especially for polymorphic functions.

Pylance (which is Microsoft's full-featured language server built on top of pyright) ships with some [additional type stubs](https://github.com/microsoft/python-type-stubs/tree/main/stubs), including partial stubs for sklearn. Have you tried using pylance?

Are you running pyright in an editor (as a language server) or on the command line? If you're running it in an editor, you can hover over an identifier to determine the evaluated type of that identifier. For example, if you see an error reported for the expression `X_train.shape` and pyright is indicating that `shape` isn't a known member of `list[Unknown]`, you can hover over `X_train` and discover that its evaluated type is `Unknown | list[Unknown]`. You can then look at where `X_train` is assigned its value. It comes from a call to `train_test_split`. You can hover over that identifier and see that its return type is `list[Unknown | list[Unknown]]`. You can then "Go To Definition" to see the code for this function and see why pyright is inferring its return type as `list[Unknown | list[Unknown]]`.

**Reply by mlauronen** [reply]
Everything is the latest version. I installed all today. I am going to dig this more later. Thank you for your support.

**Reply by mlauronen** [reply]
Uh. Forgot to say that vscode/pylance works ok.

**Reply by erictraut** [reply]
If you're seeing different results between pylance and pyright, then it's most likely because of the `sklearn` partial stubs that pylance bundles. You may find [this documentation](https://github.com/microsoft/pylance-release/blob/main/USING_WITH_PYRIGHT.md) useful.

**Comment by mlauronen** [comment, marked as answer]
Problem solved. The modules must be imported correctly. Here is an example:
```python
import numpy as np
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense, Input
```
